src/main/scala/streamer/AdressGenerationUnit.scala

```diff
@@ -0,0 +1,198 @@
+
+  // Assuming addr_gen_counter and loop_counters are of type UInt
+  // counter tracking address generation process
+  val addr_gen_counter = RegInit(
+    0.U((temporalLoopDim * temporalLoopBoundWidth).W)
+  )
+  // sub-counters for tracking index for each nested loop
+  val loop_counters = WireInit(
+    VecInit(Seq.fill(temporalLoopDim)(0.U(temporalLoopBoundWidth.W)))
+  )
+
+  // signals indicating can generating next address
+  // signal indicating next address can be generated
+  val addr_gen_counter_inc = Wire(Bool())
+
+  // State declaration
+  val sIDLE :: sBUSY :: Nil = Enum(2)
+  val cstate = RegInit(sIDLE)
+  val nstate = WireInit(sIDLE)
+
+  // Changing states
+  cstate := nstate
+
+  chisel3.dontTouch(cstate)
+  switch(cstate) {
+    is(sIDLE) {
+      when(config_valid) {
+        nstate := sBUSY
+      }.otherwise {
+        nstate := sIDLE
+      }
+
+    }
+    is(sBUSY) {
+      when(addr_gen_finish) {
+        nstate := sIDLE
+      }.otherwise {
+        nstate := sBUSY
+      }
+    }
+  }
+
+  // when the config is valid, store the configuration for address generation
+  config_valid := io.temporalLoopBounds_i.fire && io.temporalStrides_i.fire && io.ptr_i.fire
+
+  when(config_valid) {
+    temporalLoopBounds := io.temporalLoopBounds_i.bits
+    temporalStrides := io.temporalStrides_i.bits
+    ptr := io.ptr_i.bits
+  }
+
+  // generating next address
+  chisel3.dontTouch(addr_gen_counter_inc)
+  addr_gen_counter_inc := io.ptr_o.ready === 1.B && cstate === sBUSY && !addr_gen_finish
+
+  when(addr_gen_counter_inc === 1.B) {
+    addr_gen_counter := addr_gen_counter + 1.U
+  }.elsewhen(cstate === sIDLE) {
+    addr_gen_counter := 0.U
+  }.otherwise {
+    addr_gen_counter := addr_gen_counter
+  }
+
+  addr_gen_finish := addr_gen_counter === temporalLoopBounds.reduce(
+    _ * _
+  ) && cstate === sBUSY
+
+  // function for generating sub-counters for each nested loop to generate the output address
+  def genNestedLoopCounter(
+      valid: Bool,
+      temporalLoopDim: Int,
+      addr_gen_counter: UInt,
+      temporalLoopBounds: Vec[UInt]
+  ): Vec[UInt] = {
+    val loop_counters = WireInit(
+      VecInit(Seq.fill(temporalLoopDim)(0.U(temporalLoopBoundWidth.W)))
+    )
+
+    // spatially unrolling the sub loop counter computation process
+    // spatially unrolling the sub loop counter computation process
+    val addr_gen_counter_next_loop = WireInit(
+      VecInit(
+        Seq.fill(temporalLoopDim + 1)(
+          0.U((temporalLoopDim * temporalLoopBoundWidth).W)
+        )
+      )
+    )
+
+    addr_gen_counter_next_loop(0) := addr_gen_counter
+
+    // Calculate sub loop counters
+    when(valid) {
+      for (i <- 0 until temporalLoopDim) {
+        loop_counters(i) := addr_gen_counter_next_loop(i) % temporalLoopBounds(
+          i
+        )
+        addr_gen_counter_next_loop(i + 1) := addr_gen_counter_next_loop(
+          i
+        ) / temporalLoopBounds(i)
+      }
+    }
+
+    loop_counters
+  }
```

Comment:
How about something like this?

```suggestion
  def genNestedLoopCounter(
      valid: Bool,
      temporalLoopDim: Int,
      temporalLoopBounds: Vec[UInt]
  ): Vec[UInt] = {

    val loop_counters = RegInit(VecInit(Seq.fill(temporalLoopDim)(0.U(temporalLoopBoundWidth.W))))
    val loop_counters_next = WireInit(VecInit(Seq.fill(temporalLoopDim)(0.U(temporalLoopBoundWidth.W))))
    val loop_counters_valid = WireInit(VecInit(Seq.fill(temporalLoopDim)(0.U(1.W))))
    val loop_counters_last = WireInit(VecInit(Seq.fill(temporalLoopDim)(0.U(1.W))))

    for (i <- 0 until temporalLoopDim) {
      loop_counters_next(i) := loop_counters(i) + 1
      loop_counters_last(i) := loop_counters_next(i) === temporalLoopBounds(i)
    }

    loop_counters_valid(0) := valid
    for (i <- 1 until temporalLoopDim) {
      loop_counters_valid(i) := loop_counters_last(i - 1) && loop_counters_valid(i - 1)
    }

    for (i <- 0 until temporalLoopDim) {
      when(loop_counters_valid(i)) {
        loop_counters(i) := loop_counters_next(i)
      }.otherwise {
        loop_counters(i) := loop_counters(i)
      }
    }

    loop_counters
  }
```

Reply:
Initial test, not what we want, we can discuss more!! I think this is better!

Reply:
aha,
`loop_counters_next(i) := Mux(loop_counters_last(i), 0.U, loop_counters(i) + 1.U)`

Reply:
Cool!